packages/web/app/src/lib/preflight-sandbox/graphiql-plugin.tsx

```diff
@@ -0,0 +1,550 @@
+          Edit
+        </Button>
+      </div>
+      <Subtitle>
+        This script is run before each operation submitted, e.g. for automated authentication.
+      </Subtitle>
+
+      <div className="flex items-center gap-2 text-sm">
+        <Switch
+          checked={!disabled}
+          onCheckedChange={v => setDisabled(!v)}
+          className="my-4"
+          data-cy="disable-preflight-script"
+        />
+        <span className="w-6">{disabled ? 'OFF' : 'ON'}</span>
+      </div>
+
+      {!disabled && (
+        <MonacoEditor
+          height={128}
+          value={script}
+          {...monacoProps.script}
+          className={classes.monacoMini}
+          wrapperProps={{
+            ['data-cy']: 'preflight-script-editor-mini',
+          }}
+          options={{
+            ...monacoProps.script.options,
+            lineNumbers: 'off',
+            readOnly: true,
+          }}
+        />
+      )}
+
+      <Title className="mt-6 flex items-center gap-2">
+        Environment variables{' '}
+        <Badge className="text-xs" variant="outline">
+          JSON
+        </Badge>
+      </Title>
+      <Subtitle>Define variables to use in your Headers</Subtitle>
+      <MonacoEditor
+        height={128}
+        value={env}
+        onChange={setEnv}
+        {...monacoProps.env}
+        className={classes.monacoMini}
+        wrapperProps={{
+          ['data-cy']: 'env-editor-mini',
+        }}
+      />
+    </>
+  );
+}
+
+function PreflightScriptModal({
+  isOpen,
+  toggle,
+  scriptValue,
+  onScriptValueChange,
+  envValue,
+  onEnvValueChange,
+}: {
+  isOpen: boolean;
+  toggle: () => void;
+  scriptValue?: string;
+  onScriptValueChange: (value?: string) => void;
+  envValue: string;
+  onEnvValueChange: (value?: string) => void;
+}) {
+  const scriptEditorRef = useRef<editor.IStandaloneCodeEditor | null>(null);
+  const envEditorRef = useRef<editor.IStandaloneCodeEditor | null>(null);
+  const [logs, setLogs] = useState<(LogMessage | { type: 'separator' })[]>([]);
+  const consoleRef = useRef<HTMLElement>(null);
+
+  const handleScriptEditorDidMount: OnMount = useCallback(editor => {
+    scriptEditorRef.current = editor;
+  }, []);
+
+  const handleEnvEditorDidMount: OnMount = useCallback(editor => {
+    envEditorRef.current = editor;
+  }, []);
+  const handleSubmit = useCallback(() => {
+    onScriptValueChange(scriptEditorRef.current?.getValue());
+    onEnvValueChange(envEditorRef.current?.getValue());
+    toggle();
+  }, []);
+
+  const handleRunScript = useCallback(async () => {
+    clearTimeout(timerId);
+    const env = envEditorRef.current?.getValue() ?? '';
+    preflightWorker.postMessage({
+      script: scriptEditorRef.current?.getValue() ?? '',
+      environmentVariables: env ? JSON.parse(env) : {},
+    });
+    const { setEnv } = usePreflightScriptStore.getState();
+    preflightWorker.onmessage = ({ data }: MessageEvent<PreflightScriptResult>) => {
```

Comment:
It seems like two event listeners are assigned to the preflightWorker, how does this even work? Can we have one event listener that handles both scenarios. Having two is a bit confusing and messy.

Reply:
> It seems like two event listeners are assigned to the preflightWorker

not two but 1, there is reassignment of `onmessage` handler

having one implies a refactoring which I couldn't succeded from first attempt, they do different job:
- `onmessage` from preflight modal handle logs/errors/env
- `onmessage` from execute query handle `errors/env`